pyomo/core/base/units_container.py

```diff
@@ -0,0 +1,1335 @@
+
+        Parameters
+        ----------
+        expr : Pyomo expression
+           the input expression for extracting units
+
+        Returns
+        -------
+        : tuple (PyomoUnit, pint unit)
+        """
+        pyomo_unit, pint_unit = _UnitExtractionVisitor(self).walk_expression(expr=expr)
+
+        if pint_unit is not None:
+            assert pyomo_unit is not None
+            if type(pint_unit) != type(self._pint_registry.kg):
+                pint_unit = pint_unit.units
+            return (_PyomoUnit(pint_unit, self._pint_registry), pint_unit)
+        return (None, None)
+
+    def get_units(self, expr):
+        """
+        Return the Pyomo units corresponding to this expression (also performs validation
+        and will raise an exception if units are not consistent).
+
+        Parameters
+        ----------
+        expr : Pyomo expression
+            The expression containing the desired units
+
+        Returns
+        -------
+        : Pyomo unit (expression)
+           Returns the units corresponding to the expression
+
+        Raises
+        ------
+        :py:class:`pyomo.core.base.units_container.UnitsError`, :py:class:`pyomo.core.base.units_container.InconsistentUnitsError`
+
+        """
+        pyomo_unit, pint_unit = self._get_units_tuple(expr=expr)
+
+        # Currently testing the idea that a PyomoUnit can contain a pint_unit expression
+        # (instead of just a single unit). If this is OK long term, then we can clean up the
+        # visitor code to only track the pint units
+        return pyomo_unit
+
+    def check_units_consistency(self, expr, allow_exceptions=True):
+        """
+        Check the consistency of the units within an expression. IF allow_exceptions is False,
+        then this function swallows the exception and returns only True or False. Otherwise,
+        it will throw an exception if the units are inconsistent.
+
+        Parameters
+        ----------
+        expr : Pyomo expression
+            The source expression to check.
+
+        allow_exceptions: bool
+            True if you want any exceptions to be thrown, False if you only want a boolean
+            (and the exception is ignored).
+
+        Returns
+        -------
+        : bool
+           True if units are consistent, and False if not
+
+        Raises
+        ------
+        :py:class:`pyomo.core.base.units_container.UnitsError`, :py:class:`pyomo.core.base.units_container.InconsistentUnitsError`
+
+        """
+        if allow_exceptions is True:
+            pyomo_unit, pint_unit = self._get_units_tuple(expr=expr)
+            # if we make it here, then no exceptions were thrown
+            return True
+
+        # allow_exceptions is not True, therefore swallow any exceptions in a try-except
+        try:
+            pyomo_unit, pint_unit = self._get_units_tuple(expr=expr)
+        except:
+            return False
+
+        return True
```

Comment:
This method is a bit strange (and repeats the call to _get_units_tuple).  What about:
```python
try:
    pyomo_unit, pint_unit = self._get_units_tuple(expr=expr)
    return True
except: # Personally, I would also only trap specific exceptions here
    if allow_exceptions:
        raise
    return False
```

Reply:
I like it. done.